### ISSUE TITLE
Bazel 6.4.0 fails to build tensorflow-official on MacOS

### ISSUE THREAD

**Issue by meteorcloudy** [bazelbuild/bazel#20487]
Copied from internal bug b/314315783

```
[22,375 / 22,439] Linking tensorflow/libtensorflow_cc.2.16.0.dylib; 3s remote-cache
[22,376 / 22,439] [Prepa] action 'SolibSymlink _solib_darwin_arm64/_Utensorflow/libtensorflow_cc.2.16.0.dylib'
ERROR: /Volumes/BuildData/tmpfs/src/github/tensorflow/tensorflow/python/BUILD:514:27: Linking tensorflow/python/_pywrap_py_exception_registry.so failed: (Exit 1): cc_wrapper.sh failed: error executing command (from target //tensorflow/python:_pywrap_py_exception_registry.so) external/local_config_cc/cc_wrapper.sh @bazel-out/darwin_arm64-opt/bin/tensorflow/python/_pywrap_py_exception_registry.so-2.params
ld: library not found for -l:_pywrap_tensorflow_internal.so
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error in child process '/usr/bin/xcrun'. 1
ERROR: /Volumes/BuildData/tmpfs/src/github/tensorflow/tensorflow/python/BUILD:1185:27: Linking tensorflow/python/_pywrap_tfe.so failed: (Exit 1): cc_wrapper.sh failed: error executing command (from target //tensorflow/python:_pywrap_tfe.so) external/local_config_cc/cc_wrapper.sh @bazel-out/darwin_arm64-opt/bin/tensorflow/python/_pywrap_tfe.so-2.params
ld: library not found for -l:_pywrap_tensorflow_internal.so
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error in child process '/usr/bin/xcrun'. 1
ERROR: /Volumes/BuildData/tmpfs/src/github/tensorflow/tensorflow/python/BUILD:466:27: Linking tensorflow/python/_pywrap_tfcompile.so failed: (Exit 1): cc_wrapper.sh failed: error executing command (from target //tensorflow/python:_pywrap_tfcompile.so) external/local_config_cc/cc_wrapper.sh @bazel-out/darwin_arm64-opt/bin/tensorflow/python/_pywrap_tfcompile.so-2.params
ld: library not found for -l:_pywrap_tensorflow_internal.so
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error in child process '/usr/bin/xcrun'. 1
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 2558.849s, Critical Path: 86.52s
INFO: 22385 processes: 7 remote cache hit, 5128 internal, 17250 local.
FAILED: Build did NOT complete successfully
```

Possible culprit: https://github.com/bazelbuild/bazel/commit/946777ab77ec84223b02b4deb2d7d947a4ab3611

/cc @oquenchil @buildbreaker2021

**Comment by meteorcloudy** [comment]
This may also be a problem in Bazel@HEAD.

**Comment by meteorcloudy** [comment]
@bazel-io fork 6.5.0

**Comment by oquenchil** [comment]
Hello @keith and @Yannic ,

This is a small repro of the issue that Tensorflow has on macOS.
```
cc_library(name = "foo")
cc_library(name = "bar")

cc_shared_library(
    name = "foo_so",
    dynamic_deps = ["bar_so"],
    deps = ["foo"],
)

cc_shared_library(
    name = "bar_so",
    shared_lib_name = "_bar_internal.so",
    deps = ["bar"]
)
```
The problem was introduced by this change https://github.com/bazelbuild/bazel/commit/044a14cca2747aeff258fc71eaeb153c08cb34d5, which fixes the following, quoted from commit message:

> This CL also fixes the name of the symlink for a shared library which was mangled so that the original name is preserved. This is important because if a distribution artifact is linked against the library with the mangled name, even if the shared library was installed in the system and the RPATH set correctly, the dynamic loader won't find it.


The line that mattears is this:
https://github.com/bazelbuild/bazel/commit/044a14cca2747aeff258fc71eaeb153c08cb34d5#diff-369601ad017c359f03f210c8e6b98c369569671e7774210182d652303acde137R853

How we mangled:
original: `libfoo.so`
mangled before the commit: `libMANGLEDfoo.so`
mangled after the commit: `libMANGLED.so/libfoo.so`

The `soname` is computed based on the name of the artifact, see [here](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/cpp/CppLinkActionBuilder.java;drc=0aacc74776ec824cc403e5c040d97a4618c64429;l=835).

At least on Linux from what I recall as the reason for the change, we must link against a library with the name as it will be found on the system (again see commit description). It could be that this only mattered if the `soname` wasn't passed, I can't remember the exact details.

Then the next part to understand the problem is the following. From the `ld` [man page for Linux](https://linux.die.net/man/1/ld), we can see that if a library that we are linking doesn't have the usual prefix and suffix we just pass `-l:filename` and `filename` will be searched for literally. I don't see that on the `ld` [man page for macOS](https://www.unix.com/man-page/macOS/1/ld/).

Before the commit, this wasn't a problem because the mangled name had the prefix and suffix. But now we are passing `-l:_bar_internal.so` on macOS and it seems `ld` is trying to find `":_bar_internal.so"`. The first question is why are the macOS toolchains passing `"-l:filename"` if that functionality doesn't exist? But then the main problem is how to move on from here because if I remove the `preserveName=true` fix for macOS, I believe the distribution artifacts that have linked against mangled symlinks will be wrong or does macOS not have the same problem?

Should I revert `preserveName=true` for macOS assuming that if `install_name` is passed then it will work correctly?

**Comment by oquenchil** [comment]
Preferably though, this is fixed in the toolchain so that I don't have to hardcode platform specific logic in Java. @hlopko mentioned that passing the full path of the dynamic library without `-l` should work. Wdyt? Is that acceptable? Could someone more familiar with macOS toolchains send the PR?

Thanks!

**Comment by iancha1992** [comment]
@bazel-io flag 6.5.0

**Comment by oquenchil** [comment]
@keith Could you please chime in? Or who else from the community might know about Apple toolchains?

**Comment by oquenchil** [comment]
@brentleyjones maybe you could take a look? The tldr is that Apple toolchains are adding `-l:libname` when the dash `l` colon syntax for specific file names for the prefix and suffix not to be added doesn't exist on macOS. Probably the right solution is to place the whole path to the file in the command line in those cases with no option preceding it (i.e. no `-l`). For an archive it would be treated as whole-archive but for a dynamic library this might be ok.

**Comment by keith** [comment]
I think using the full path sounds reasonable if that works. Is the repo as simple as taking tensorflow @ HEAD, forcing bazel 6.4.0 and doing `bazel build //tensorflow/tools/pip_package:build_pip_package`?

**Comment by keith** [comment]
> For an archive it would be treated as whole-archive but for a dynamic library this might be ok.

In a quick test outside of bazel there is no difference in doing `-l foo` vs passing `libfoo.dylib` directly on the linker command line even if the contents of the dylib are entirely unused. In either case the resulting binary ends up with a dependency on the dylib, so I don't think there is a difference for this case.

**Comment by keith** [comment]
It looks to me like in the crosstool there isn't a way to get the full path to the dynamic lib. Right now it's using `libraries_to_link.name`, but `name` in `path/to/foo.dylib` is `foo`, and to pass it directly we need that full path. AFAICT these are the only potential fields on this https://github.com/keith/bazel/blob/f8257b66ab6d9e2f6c47fdd7ff320f47781de4ea/src/main/java/com/google/devtools/build/lib/rules/cpp/CcToolchainVariables.java#L662-L665

Am I understanding that correctly? (I can repro locally and am testing with the apple_support toolchain even though right now tensorflow is using the builtin toolchain with 6.4.0 just to reduce the iteration cycle)

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 6.5.0 RC1](https://releases.bazel.build/6.5.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
